### PR TITLE
Disabling ASM mode if the Istio CRD is not available.

### PR DESCRIPTION
**Pull request by cadmuxe**

### CODE REVIEW

File: pkg/flags/flags.go [hunk 1]
```diff
@@ -204,7 +204,7 @@ L7 load balancing. CSV values accepted. Example: -node-port-ranges=80,8080,400-5
 	flag.BoolVar(&F.EnableL7Ilb, "enable-l7-ilb", false,
 		`Optional, whether or not to enable L7-ILB.`)
 	flag.BoolVar(&F.EnableASMConfigMapBasedConfig, "enable-asm-config-map-config", false, "Enable ASMConfigMapBasedConfig")
-	flag.StringVar(&F.ASMConfigMapBasedConfigNamespace, "asm-configmap-based-config-namespace", "kube-system,istio-system", "ASM Configmap based config: configmap namespace")
```
Comment (cadmuxe) [inline pkg/flags/flags.go]: This is a bug, it's the configmap namespace, but not the skip namespace.

File: pkg/context/context.go [hunk 2]
```diff
@@ -135,25 +139,63 @@ func (ctx *ControllerContext) Init() {
...
+		destinationRuleGroup      = "networking.istio.io"
+		destinationRuleAPIVersion = "v1alpha3"
+		destinationRulePlural     = "destinationrules"
+		// This must match the spec fields below, and be in the form: <plural>.<group>
+		destinationRuleCRDName = "destinationrules.networking.istio.io"
+	)
+
+	apiextensionClient, err := apiextensionsclientset.NewForConfig(ctx.KubeConfig)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to validate DestinationRule CRD: failed to create apiextensionClient, disabling ASM Mode, error: %s", err)
```
Comment (freehan) [inline pkg/context/context.go]: I am not sure if this is related to DestinationRule CRD at all. This just means that the clientSet cannot be initialized, right?

File: pkg/context/context.go [hunk 3]
```diff
@@ -135,25 +139,63 @@ func (ctx *ControllerContext) Init() {
...
+	if err != nil {
+		msg := fmt.Sprintf("Failed to create kubernetes dynamic client, disabling ASM Mode, please retry. Error: %v", err)
+		klog.Fatalf(msg)
+		ctx.ASMConfigController.RecordEvent("Warning", "FailedCreateDynamicClient", msg)
+		ctx.ASMConfigController.DisableASMMode()
+		return
+	}
+
+	klog.Info("The DestinationRule group version is v1alpha3 in group networking.istio.io. Need to update as istio API graduates.")
+	destrinationGVR := schema.GroupVersionResource{Group: "networking.istio.io", Version: "v1alpha3", Resource: "destinationrules"}
```
Comment (freehan) [inline pkg/context/context.go]: use constant vars above?

File: pkg/context/context.go [hunk 4]
```diff
@@ -135,25 +139,63 @@ func (ctx *ControllerContext) Init() {
...
+	dynamicClient, err := dynamic.NewForConfig(ctx.KubeConfig)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to create kubernetes dynamic client, disabling ASM Mode, please retry. Error: %v", err)
+		klog.Fatalf(msg)
+		ctx.ASMConfigController.RecordEvent("Warning", "FailedCreateDynamicClient", msg)
+		ctx.ASMConfigController.DisableASMMode()
+		return
+	}
+
+	klog.Info("The DestinationRule group version is v1alpha3 in group networking.istio.io. Need to update as istio API graduates.")
```
Comment (freehan) [inline pkg/context/context.go]: V(2)

use constant vars above?

File: pkg/cmconfig/config.go [hunk 5]
```diff
@@ -24,7 +24,7 @@ const (
 
 // NewConfig returns a Conifg instances with default values.
 func NewConfig() Config {
-	return Config{ASMServiceNEGSkipNamespaces: []string{"kube-system"}}
+	return Config{ASMServiceNEGSkipNamespaces: []string{"kube-system", "istio-system"}}
```
[the remark this answers is not in the document]
Comment (cadmuxe) [reply pkg/cmconfig/config.go]: istio-system was in the default skip flag, I just forgot to move it here.

File: pkg/context/context.go (hunk 2 above)
Comment (cadmuxe) [reply pkg/context/context.go]: yes, but it's a required step to validate the CRD

File: pkg/cmconfig/config.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (cadmuxe) [reply pkg/cmconfig/config.go]: yes, explicitly config is good.

File: pkg/cmconfig/controller.go [hunk 6]
```diff
@@ -58,6 +63,23 @@ func (c *ConfigMapConfigController) GetConfig() Config {
 	return *c.currentConfig
 }
 
+// DisableASMMode disables the ASM Mode by updating the ConfigMap and setting the internal flag.
+func (c *ConfigMapConfigController) DisableASMMode() {
```
Comment (freehan) [inline pkg/cmconfig/controller.go]: I am thinking if it is possible to not patch the original config map but just record events. Patching configmap requires some RBAC changes I assume?

File: pkg/context/context.go (hunk 2 above)
Comment (freehan) [reply pkg/context/context.go]: I mean the message should be failed to create extension clientset right? Not validating destinationRule CRD.

**Pull request merged**